components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +87,166 @@ public AbstractHTTPFunction() {
+                } else {
+                    request = clientRequest;
+                }
 
-            if (!isValidRequestDomain(request.getURI())) {
-                outcome = Constants.OUTCOME_FAIL;
-                LOG.error("Provided Url does not contain a allowed domain. Invalid Url: " + endpointURL);
-                asyncReturn.accept(context, Collections.emptyMap(), outcome);
-                return;
-            }
+                if (request.getURI() != null) {
+                    endpointURL = request.getURI().toString();
+                }
 
-            try (CloseableHttpResponse response = client.execute(request)) {
-                responseCode = response.getStatusLine().getStatusCode();
-                if (responseCode >= 200 && responseCode < 300) {
-                    outcome = Constants.OUTCOME_SUCCESS;
-                    if (response.getEntity() != null) {
-                        Header contentType = response.getEntity().getContentType();
-                        String jsonString = EntityUtils.toString(response.getEntity());
-                        if (contentType != null && contentType.getValue().contains(TYPE_TEXT_PLAIN)) {
-                            // For 'text/plain', put the response body into the JSON object as a single field.
-                            json = new JSONObject();
-                            json.put(RESPONSE, jsonString);
-                        } else {
-                            JSONParser parser = new JSONParser();
-                            json = (JSONObject) parser.parse(jsonString);
+                if (!isValidRequestDomain(request.getURI())) {
+                    LOG.error("Provided Url does not contain a allowed domain. Invalid Url: " + endpointURL);
+                    asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+                } else {
+                    Pair<String, JSONObject> result = executeRequest(request, endpointURL);
+                    if (result.getLeft().equals(Constants.OUTCOME_FAIL)) {
+                        if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                            DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                                    DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                                    Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                            diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.EXTERNAL_API, endpointURL)
+                                    .resultMessage("Error while calling endpoint, proceeding with retry attempts.")
+                                    .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                                    .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                            LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
```

Comment:
Above is bit too generic for the debug log.

We should be able to identify if the request timed out or the received response is an error state (> 500).
Also, if the call is 401 we don't need a retry as it just says it's unauthorized so we just need to add that to the log.

I don't think you can log with above context here. Either have to modify the function or move the debugger there

Reply:
Fixed with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/05c4f77e8d18ba911c3b98bc7e18ec3fb73dcd33